internal/command/command_service.go

```diff
@@ -0,0 +1,264 @@
+	}
+}
+
+func (cs *CommandService) UpdateDataPlaneStatus(ctx context.Context, resource *mpi.Resource) error {
+	if !cs.isConnected.Load() {
+		err := cs.createConnection(ctx, resource)
+		if err != nil {
+			return err
+		}
+
+		var subscribeCtx context.Context
+
+		cs.subscribeMutex.Lock()
+		subscribeCtx, cs.subscribeCancel = context.WithCancel(ctx)
+		cs.subscribeMutex.Unlock()
+
+		go cs.subscribe(subscribeCtx)
+	}
+
+	correlationID := logger.GetCorrelationID(ctx)
+
+	request := &mpi.UpdateDataPlaneStatusRequest{
+		MessageMeta: &mpi.MessageMeta{
+			MessageId:     uuid.NewString(),
+			CorrelationId: correlationID,
+			Timestamp:     timestamppb.Now(),
+		},
+		Resource: resource,
+	}
+
+	backOffCtx, backoffCancel := context.WithTimeout(ctx, cs.agentConfig.Common.MaxElapsedTime)
+	defer backoffCancel()
+
+	sendDataPlaneStatus := func() (*mpi.UpdateDataPlaneStatusResponse, error) {
+		slog.DebugContext(ctx, "Sending data plane status update request", "request", request)
+		if cs.commandServiceClient == nil {
+			return nil, errors.New("command service client is not initialized")
+		}
+
+		response, updateError := cs.commandServiceClient.UpdateDataPlaneStatus(ctx, request)
+
+		validatedError := grpc.ValidateGrpcError(updateError)
+		if validatedError != nil {
+			slog.ErrorContext(ctx, "Failed to send update data plane status", "error", validatedError)
+
+			return nil, validatedError
+		}
+
+		return response, nil
+	}
+
+	response, err := backoff.RetryWithData(
+		sendDataPlaneStatus,
+		backoffHelpers.Context(backOffCtx, cs.agentConfig.Common),
+	)
+	if err != nil {
+		return err
+	}
+	slog.DebugContext(ctx, "UpdateDataPlaneStatus response", "response", response)
+
+	return err
+}
+
+func (cs *CommandService) UpdateDataPlaneHealth(ctx context.Context, instanceHealths []*mpi.InstanceHealth) error {
+	if !cs.isConnected.Load() {
+		return errors.New("command service client not connected yet")
+	}
+
+	correlationID := logger.GetCorrelationID(ctx)
+
+	request := &mpi.UpdateDataPlaneHealthRequest{
+		MessageMeta: &mpi.MessageMeta{
+			MessageId:     uuid.NewString(),
+			CorrelationId: correlationID,
+			Timestamp:     timestamppb.Now(),
+		},
+		InstanceHealths: instanceHealths,
+	}
+
+	backOffCtx, backoffCancel := context.WithTimeout(ctx, cs.agentConfig.Common.MaxElapsedTime)
+	defer backoffCancel()
+
+	sendDataPlaneHealth := func() (*mpi.UpdateDataPlaneHealthResponse, error) {
+		slog.DebugContext(ctx, "Sending data plane health update request", "request", request)
+		if cs.commandServiceClient == nil {
+			return nil, errors.New("command service client is not initialized")
+		}
+
+		response, updateError := cs.commandServiceClient.UpdateDataPlaneHealth(ctx, request)
+
+		validatedError := grpc.ValidateGrpcError(updateError)
+
+		if validatedError != nil {
+			slog.ErrorContext(ctx, "Failed to send update data plane health", "error", validatedError)
+
+			return nil, validatedError
+		}
+
+		return response, nil
+	}
```

Comment:
I wonder if we could parameterize these `func`s via generics or something, as we seem to be repeating this same pattern.

Reply:
I did try to take a look at how we could reduce the similar code but I couldn't see an easy way of doing it.